Description: Open PHC device for writing with extpps option
 In version 6.15 the Linux kernel started checking write access on the
 PHC file descriptor in the PTP_PIN_SETFUNC and PTP_EXTTS_REQUEST ioctls.
 chronyd opened the PHC device as readonly, which caused the PHC refclock
 driver configured with the extpps option to fail with the
 "Could not enable external PHC timestamping" error message.

 To ensure compatibility with new kernel versions, add flags to the
 SYS_Linux_OpenPHC() function and open the device with the O_RDWR flag
 when the extpps option is enabled.
Origin: backport, https://gitlab.com/chrony/chrony/-/commit/f78e4681eff71d941fab3be5ee406d920a155a20
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1127659
Last-Update: 2026-02-12
---
Index: chrony/ntp_io_linux.c
===================================================================
--- chrony.orig/ntp_io_linux.c
+++ chrony/ntp_io_linux.c
@@ -220,7 +220,7 @@ add_interface(CNF_HwTsInterface *conf_if
 
   SCK_CloseSocket(sock_fd);
 
-  phc_fd = SYS_Linux_OpenPHC(NULL, ts_info.phc_index);
+  phc_fd = SYS_Linux_OpenPHC(NULL, ts_info.phc_index, O_RDONLY);
   if (phc_fd < 0)
     return 0;
 
Index: chrony/refclock_phc.c
===================================================================
--- chrony.orig/refclock_phc.c
+++ chrony/refclock_phc.c
@@ -66,7 +66,7 @@ static int phc_initialise(RCL_Instance i
 {
   const char *options[] = {"nocrossts", "extpps", "pin", "channel", "clear", NULL};
   struct phc_instance *phc;
-  int phc_fd, rising_edge;
+  int rising_edge;
   struct stat st;
   char *path, *s;
 
@@ -74,19 +74,20 @@ static int phc_initialise(RCL_Instance i
 
   path = RCL_GetDriverParameter(instance);
  
-  phc_fd = SYS_Linux_OpenPHC(path, 0);
-  if (phc_fd < 0)
-    LOG_FATAL("Could not open PHC");
-
   phc = MallocNew(struct phc_instance);
-  phc->fd = phc_fd;
-  if (fstat(phc_fd, &st) < 0 || !S_ISCHR(st.st_mode))
-    LOG_FATAL("Could not get PHC index");
-  phc->dev_index = minor(st.st_rdev);
   phc->mode = 0;
   phc->nocrossts = RCL_GetDriverOption(instance, "nocrossts") ? 1 : 0;
   phc->extpps = RCL_GetDriverOption(instance, "extpps") ? 1 : 0;
   UTI_ZeroTimespec(&phc->last_extts);
+
+  phc->fd = SYS_Linux_OpenPHC(path, 0, phc->extpps ? O_RDWR : O_RDONLY);
+  if (phc->fd < 0)
+    LOG_FATAL("Could not open PHC");
+
+  if (fstat(phc->fd, &st) < 0 || !S_ISCHR(st.st_mode))
+    LOG_FATAL("Could not get PHC index");
+  phc->dev_index = minor(st.st_rdev);
+
   phc->clock = HCL_CreateInstance(0, 16, UTI_Log2ToDouble(RCL_GetDriverPoll(instance)),
                                   RCL_GetPrecision(instance));
 
Index: chrony/sys_linux.c
===================================================================
--- chrony.orig/sys_linux.c
+++ chrony/sys_linux.c
@@ -903,7 +903,7 @@ get_precise_phc_readings(int phc_fd, int
 /* ================================================== */
 
 int
-SYS_Linux_OpenPHC(const char *path, int phc_index)
+SYS_Linux_OpenPHC(const char *path, int phc_index, int flags)
 {
   struct ptp_clock_caps caps;
   char phc_path[64];
@@ -915,7 +915,7 @@ SYS_Linux_OpenPHC(const char *path, int
     path = phc_path;
   }
 
-  phc_fd = open(path, O_RDONLY);
+  phc_fd = open(path, flags);
   if (phc_fd < 0) {
     LOG(LOGS_ERR, "Could not open %s : %s", path, strerror(errno));
     return -1;
Index: chrony/sys_linux.h
===================================================================
--- chrony.orig/sys_linux.h
+++ chrony/sys_linux.h
@@ -39,7 +39,7 @@ extern void SYS_Linux_EnableSystemCallFi
 
 extern int SYS_Linux_CheckKernelVersion(int req_major, int req_minor);
 
-extern int SYS_Linux_OpenPHC(const char *path, int phc_index);
+extern int SYS_Linux_OpenPHC(const char *path, int phc_index, int flags);
 
 extern int SYS_Linux_GetPHCReadings(int fd, int nocrossts, int *reading_mode, int max_readings,
                                     struct timespec tss[][3]);
